linux_os/guide/system/accounts/accounts-restrictions/password_storage/no_empty_passwords/ansible/shared.yml

```diff
@@ -1,14 +1,72 @@
-# platform = multi_platform_rhel,multi_platform_fedora,multi_platform_ol,multi_platform_rhv,multi_platform_sle
+# platform = multi_platform_rhel,multi_platform_fedora,multi_platform_ol,multi_platform_rhv
 # reboot = false
 # strategy = configure
 # complexity = low
 # disruption = medium
-- name: "Prevent Log In to Accounts With Empty Password - system-auth"
-  replace:
-    dest: /etc/pam.d/system-auth
-    regexp: 'nullok'
+- name: Check if system relies on authselect
+  ansible.builtin.stat:
+    path: /usr/bin/authselect
+  register: result_authselect_present
+
+- name: Check integrity of authselect current profile
+  ansible.builtin.command:
+    cmd: authselect check
+  register: result_authselect_check_cmd
+  ignore_errors: yes
+  when:
+    - result_authselect_present.stat.exists
+
+- name: Get authselect current profile
+  ansible.builtin.shell:
+    cmd: authselect current -r | awk '{ print $1 }'
+  register: result_authselect_profile
+  when:
+    - result_authselect_check_cmd is failed
+
+- name: Get authselect current features
+  ansible.builtin.shell:
+    cmd: authselect current | tail -n+3 | awk '{ print $2 }'
+  register: result_authselect_features
+  when:
+    - result_authselect_check_cmd is failed
+
+- name: Define the current authselect profile as a local fact
+  ansible.builtin.set_fact:
+    authselect_profile: "{{ result_authselect_profile.stdout }}"
+  when:
+    - result_authselect_check_cmd is failed
+    - result_authselect_profile | default(false) | bool
+
+- name: Restore the authselect profile if manually modified
```

Comment:
Is it a good approach to remove any changes if the pam file was changed by a third party?

I think it makes sense for our remediations not to make direct edits when `authselect` is available in the system.
But if a change was made by something else, we should probably:
-  not make any changes, or
- fall back to the direct edit too.

Thoughts?

Reply:
Yes, this is a good approach in my point of view.

Doing manual changes when the `authselect` is present makes the pam files inconsistent. It means any other integration or even new changes can no longer happen automatically. It breaks the pam structure in a way that only manual changes are possible, which may bring at least two risks related to:
- Availability, considering how sensible the pam files are to ordering and syntax.
- Malicious activity, since manual editing of these files could be maliciously prepared to workaround the controls in place.

This remediation turns the system back to the consistent state preserving possible features previously enabled and also ensuring a validated configuration is in use.

Indeed, if any change was make directly in pam files, going against the best practices, they will be lost (for good reasons).

This specific rule deals with `nullok` parameter, which has a specific feature out-of-box in `authselect`. For other pam related rules (which I am already fixing too), the approach in cases `authselect` doesn't have a feature out-of-box, is to work with custom profiles managed by `authselect`.

Reply:
> Yes, this is a good approach in my point of view.
> 
> Doing manual changes when the `authselect` is present makes the pam files inconsistent. It means any other integration or even new changes can no longer happen automatically. It breaks the pam structure in a way that only manual changes are possible, which may bring at least two risks related to:
> 
>     * Availability, considering how sensible the pam files are to ordering and syntax.
> 
>     * Malicious activity, since manual editing of these files could be maliciously prepared to workaround the controls in place.

When one edits the PAM files without `authselect` one assumes the risks you mention. Maybe this should be more clear in the rule description? (And BTW `authselect` is not mentioned at all in the rule description)

> This remediation turns the system back to the consistent state preserving possible features previously enabled and also ensuring a validated configuration is in use.
> 
> Indeed, if any change was make directly in pam files, going against the best practices, they will be lost (for good reasons).

I'm not convinced that reverting third-party changes is a good experience, we should at least mention this in rule description.

What I think the remediation should do is warn about inconsistent state and recommend the admin to include them in a custom profile. In other words, if an admin is not using `authselect` we should not force them, but invite them to use it.

> 
> This specific rule deals with `nullok` parameter, which has a specific feature out-of-box in `authselect`. For other pam related rules (which I am already fixing too), the approach in cases `authselect` doesn't have a feature out-of-box, is to work with custom profiles managed by `authselect`.

The change in the PAM files may be related to something else, not necessarily `nullok`.

Reply:
> But if a change was made by something else, we should probably:
>
>    not make any changes, or
>    fall back to the direct edit too.

One could argue that we can just do direct edits when there are changes outside of `authselect`, because we need to maintain this approach for RHEL7 anyway.  The downside is that the rule will be remediated and the admin will not be nudged to use `authselect`.
So just I lean towards not making any changes.

Reply:
> > Yes, this is a good approach in my point of view.
> > Doing manual changes when the `authselect` is present makes the pam files inconsistent. It means any other integration or even new changes can no longer happen automatically. It breaks the pam structure in a way that only manual changes are possible, which may bring at least two risks related to:
> > ```
> > * Availability, considering how sensible the pam files are to ordering and syntax.
> > 
> > * Malicious activity, since manual editing of these files could be maliciously prepared to workaround the controls in place.
> > ```
> 
> When one edits the PAM files without `authselect` one assumes the risks you mention. Maybe this should be more clear in the rule description? (And BTW `authselect` is not mentioned at all in the rule description)

I agree that this should be more explicit here.
> 
> > This remediation turns the system back to the consistent state preserving possible features previously enabled and also ensuring a validated configuration is in use.
> > Indeed, if any change was make directly in pam files, going against the best practices, they will be lost (for good reasons).
> 
> I'm not convinced that reverting third-party changes is a good experience, we should at least mention this in rule description.
> 
I can't remember an example of current integrations with deals directly in pam files for distros which relies on `authselect`. The integrations usually happen directly on sssd layer.

> What I think the remediation should do is warn about inconsistent state and recommend the admin to include them in a custom profile. In other words, if an admin is not using `authselect` we should not force them, but invite them to use it.
> 
> > This specific rule deals with `nullok` parameter, which has a specific feature out-of-box in `authselect`. For other pam related rules (which I am already fixing too), the approach in cases `authselect` doesn't have a feature out-of-box, is to work with custom profiles managed by `authselect`.
> 
> The change in the PAM files may be related to something else, not necessarily `nullok`.

That's what I meant. : ) In other words, where the `authselect` has a feature out-of-box, we can use it in the current profile. Otherwise, a custom profile should be used, like in https://github.com/ComplianceAsCode/content/pull/8030